File: src/commands.c
#include "commands.h"
#include "ansi.h"
#include "ui.h"
#include "session.h"
#include "util.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

const orc_cmd orc_cmds[] = {
    {"/model", "[slug]", "set or show the model"},
    {"/effort", "low|medium|high", "set reasoning effort"},
    {"/new", "", "start a fresh session"},
    {"/help", "", "list commands"},
    {"/quit", "", "exit orc"},
    {NULL, NULL, NULL},
};

/* cfg->model / cfg->effort normally point at argv or string literals; /model
 * and /effort repoint them here. */
static char model_buf[128], effort_buf[16];

static const provider *cur_prov;
static const orc_cfg *cur_cfg;
static cJSON *models_cache;

void commands_init(const provider *prov, const orc_cfg *cfg) {
    cur_prov = prov;
    cur_cfg = cfg;
}

const char *commands_current_model(void) {
    return cur_cfg && cur_cfg->model ? cur_cfg->model : "";
}

cJSON *commands_models(void) {
    if (!models_cache && cur_prov && cur_prov->models)
        models_cache = cur_prov->models();
    return models_cache;
}

static const char *model_field(cJSON *m, const char *key) {
    cJSON *v = cJSON_GetObjectItem(m, key);
    return cJSON_IsString(v) ? v->valuestring : "";
}

static long long ctx_used; /* tokens in the context after the last request */

/* Context window of the current model, 0 when unknown. */
static long long model_ctx_window(void) {
    cJSON *m;
    cJSON_ArrayForEach(m, commands_models()) {
        if (strcmp(model_field(m, "slug"), commands_current_model()) != 0)
            continue;
        cJSON *w = cJSON_GetObjectItem(m, "context_window");
        return cJSON_IsNumber(w) ? (long long)w->valuedouble : 0;
    }
    return 0;
}

/* Branch name (or short detached SHA) from the nearest .git/HEAD upward. */
static void git_branch(char *out, size_t outsz) {
    out[0] = '\0';
    char dir[1024];
    if (!getcwd(dir, sizeof dir)) return;
    for (;;) {
        char path[1200];
        snprintf(path, sizeof path, "%s/.git/HEAD", dir);
        char *head = read_file(path, NULL);
        if (head) {
            head[strcspn(head, "\n")] = '\0';
            if (strncmp(head, "ref: refs/heads/", 16) == 0)
                snprintf(out, outsz, "%s", head + 16);
            else
                snprintf(out, outsz, "%.8s", head);
            free(head);
            return;
        }
        char *slash = strrchr(dir, '/');
        if (!slash || slash == dir) return;
        *slash = '\0';
    }
}

void commands_status_update(void) {
    if (!cur_cfg) return;
    char cwd[1024] = "", branch[128];
    if (!getcwd(cwd, sizeof cwd)) cwd[0] = '\0';
    const char *base = strrchr(cwd, '/');
    base = base && base[1] ? base + 1 : cwd;
    git_branch(branch, sizeof branch);

    char s[256];
    int n = snprintf(s, sizeof s, "%s · %s · %s",
                     cur_cfg->model ? cur_cfg->model : "?", cur_cfg->effort, base);
    if (branch[0] && n > 0 && n < (int)sizeof s)
        n += snprintf(s + n, sizeof s - (size_t)n, " (%s)", branch);
    long long win = model_ctx_window();
    if (ctx_used > 0 && n > 0 && n < (int)sizeof s) {
        n += ctx_used < 10000
                 ? snprintf(s + n, sizeof s - (size_t)n, " · ctx %.1fk",
                            (double)ctx_used / 1000.0)
                 : snprintf(s + n, sizeof s - (size_t)n, " · ctx %lldk",
                            ctx_used / 1000);
        if (win > 0 && n > 0 && n < (int)sizeof s)
            snprintf(s + n, sizeof s - (size_t)n, " (%lld%%)",
                     (ctx_used * 100 + win - 1) / win); /* ceil: never 0% */
    }
    ui_status_set(s);
}

void commands_ctx_used(long long tokens) {
    ctx_used = tokens;
    commands_status_update();
}

static void cmd_new(agent *ag) {
    orc_session *s = ag->sess;
    session_close(s);
    if (s->items == 0) unlink(s->path); /* nothing was said; drop the file */
    s->items = 0;
    cJSON_Delete(ag->history);
    ag->history = cJSON_CreateArray();
    uuid4(ag->cfg->session_id);
    if (session_new(s, ag->cfg) != 0) {
        fprintf(stderr, "❌ orc: cannot create session file\n");
        return;
    }
    printf("✨ new session %.8s\n", ag->cfg->session_id);
    commands_ctx_used(0); /* empty history: reset the context gauge */
}

int command_dispatch(agent *ag, const char *line) {
    if (line[0] != '/') return 0;
    size_t n = strcspn(line, " \t");
    const char *arg = line + n;
    while (*arg == ' ' || *arg == '\t') arg++;

    const orc_cmd *cmd = NULL;
    for (const orc_cmd *c = orc_cmds; c->name; c++)
        if (strlen(c->name) == n && strncmp(line, c->name, n) == 0) cmd = c;
    if (!cmd) {
        /* "/tmp/x ..." is a path, not a typo'd command; hand it to the model */
        if (memchr(line + 1, '/', n - 1)) return 0;
        printf("⚠️  unknown command %.*s (try /help)\n", (int)n, line);
        return 1;
    }

    if (strcmp(cmd->name, "/quit") == 0) return 2;
    if (strcmp(cmd->name, "/help") == 0) {
        int tty = isatty(1);
        for (const orc_cmd *c = orc_cmds; c->name; c++)
            printf(tty ? "  " BOLD("%-8s") " %-16s " DIM("%s") "\n"
                       : "  %-8s %-16s %s\n",
                   c->name, c->args, c->desc);
        return 1;
    }
    if (strcmp(cmd->name, "/new") == 0) {
        cmd_new(ag);
        return 1;
    }
    if (strcmp(cmd->name, "/model") == 0) {
        cJSON *models = commands_models();
        cJSON *m;
        if (!*arg) {
            printf("🤖 model %s (%s effort)\n", ag->cfg->model, ag->cfg->effort);
            int tty = isatty(1);
            cJSON_ArrayForEach(m, models) {
                const char *slug = model_field(m, "slug");
                int cur = strcmp(slug, ag->cfg->model) == 0;
                printf("  %s%c %-22s %s%s\n", cur && tty ? ANSI_BOLD : "",
                       cur ? '*' : ' ', slug, model_field(m, "description"),
                       cur && tty ? ANSI_RESET : "");
            }
            return 1;
        }
        int known = models == NULL; /* no list: trust the user */
        cJSON_ArrayForEach(m, models)
            if (strcmp(model_field(m, "slug"), arg) == 0) known = 1;
        snprintf(model_buf, sizeof model_buf, "%s", arg);
        ag->cfg->model = model_buf;
        printf("✅ model set to %s%s\n", model_buf,
               known ? "" : " (not in provider's model list)");
        session_set_cfg(ag->sess, ag->cfg);
        commands_status_update();
        return 1;
    }
    if (strcmp(cmd->name, "/effort") == 0) {
        if (!*arg) {
            printf("🧠 effort %s\n", ag->cfg->effort);
            return 1;
        }
        if (strcmp(arg, "low") != 0 && strcmp(arg, "medium") != 0 &&
            strcmp(arg, "high") != 0) {
            printf("⚠️  effort must be low, medium, or high\n");
            return 1;
        }
        snprintf(effort_buf, sizeof effort_buf, "%s", arg);
        ag->cfg->effort = effort_buf;
        printf("✅ effort set to %s\n", effort_buf);
        session_set_cfg(ag->sess, ag->cfg);
        commands_status_update();
        return 1;
    }
    return 1;
}

File: internal/commands/commands.go
// Package commands implements slash commands and the status line.
package commands

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/brijbyte/orc/internal/agent"
	"github.com/brijbyte/orc/internal/config"
	"github.com/brijbyte/orc/internal/provider"
	"github.com/brijbyte/orc/internal/session"
	"github.com/google/uuid"
)

type Cmd struct{ Name, Args, Desc string }

var Cmds = []Cmd{
	{"/model", "[slug]", "set or show the model"},
	{"/effort", "low|medium|high", "set reasoning effort"},
	{"/new", "", "start a fresh session"},
	{"/compact", "", "summarize history into a fresh context"},
	{"/resume", "[id]", "switch to another session"},
	{"/status", "", "show session status"},
	{"/login", "", "sign in to the provider (browser OAuth)"},
	{"/help", "", "list commands"},
	{"/quit", "", "exit orc"},
}

// UI receives command output (whole lines) and status text.
type UI interface {
	Printf(format string, a ...any)
	SetStatus(s string)
}

type Commands struct {
	prov         provider.Provider
	cfg          *config.Config
	ui           UI
	models       []provider.Model
	modelsLoaded bool
	ctxUsed      int64
	customOnce   sync.Once
	customList   []CustomCmd
}

func New(prov provider.Provider, cfg *config.Config, ui UI) *Commands {
	return &Commands{prov: prov, cfg: cfg, ui: ui}
}

func (c *Commands) CurrentModel() string { return c.cfg.Model }

func (c *Commands) Models() []provider.Model {
	if !c.modelsLoaded {
		c.models = c.prov.Models()
		c.modelsLoaded = true
	}
	return c.models
}

// modelCtxWindow is the context window of the current model, 0 when unknown.
func (c *Commands) modelCtxWindow() int64 {
	for _, m := range c.Models() {
		if m.Slug == c.cfg.Model {
			return m.ContextWindow
		}
	}
	return 0
}

// gitBranch finds the branch name (or short detached SHA) from the nearest
// .git/HEAD upward.
func gitBranch(dir string) string {
	if dir == "" {
		var err error
		if dir, err = os.Getwd(); err != nil {
			return ""
		}
	}
	for {
		head, err := os.ReadFile(filepath.Join(dir, ".git/HEAD"))
		if err == nil {
			line, _, _ := strings.Cut(string(head), "\n")
			if ref, ok := strings.CutPrefix(line, "ref: refs/heads/"); ok {
				return ref
			}
			if len(line) > 8 {
				return line[:8]
			}
			return line
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return ""
		}
		dir = parent
	}
}

func (c *Commands) StatusUpdate() {
	cwd := c.cfg.Cwd
	if cwd == "" {
		cwd, _ = os.Getwd()
	}
	base := filepath.Base(cwd)
	s := fmt.Sprintf("%s · %s · %s", c.cfg.Model, c.cfg.Effort, base)
	if branch := gitBranch(cwd); branch != "" {
		s += fmt.Sprintf(" (%s)", branch)
	}
	if c.ctxUsed > 0 {
		if c.ctxUsed < 10000 {
			s += fmt.Sprintf(" · ctx %.1fk", float64(c.ctxUsed)/1000.0)
		} else {
			s += fmt.Sprintf(" · ctx %dk", c.ctxUsed/1000)
		}
		if win := c.modelCtxWindow(); win > 0 {
			s += fmt.Sprintf(" (%d%%)", (c.ctxUsed*100+win-1)/win) // ceil: never 0%
		}
	}
	c.ui.SetStatus(s)
}

func (c *Commands) CtxUsed(tokens int64) {
	c.ctxUsed = tokens
	c.StatusUpdate()
}

func (c *Commands) cmdNew(ag *agent.Agent) {
	s := ag.Sess
	s.Close()
	if s.Items == 0 {
		os.Remove(s.Path) // nothing was said; drop the file
	}
	ag.History = nil
	c.cfg.SessionID = uuid.NewString()
	next, err := session.New(c.cfg)
	if err != nil {
		c.ui.Printf("❌ orc: cannot create session file")
		return
	}
	*ag.Sess = *next
	c.ui.Printf("✨ new session %.8s", c.cfg.SessionID)
	c.CtxUsed(0) // empty history: reset the context gauge
}

// Dispatch handles a slash command. Returns handled and quit; a non-empty
// prompt (from a custom command) must be run as a model turn by the caller.
func (c *Commands) Dispatch(ag *agent.Agent, line string) (handled, quit bool, prompt string) {
	if !strings.HasPrefix(line, "/") {
		return false, false, ""
	}
	word, rest, _ := strings.Cut(line, " ")
	if w, r, found := strings.Cut(word, "\t"); found {
		word, rest = w, r+" "+rest
	}
	arg := strings.TrimLeft(rest, " \t")

	var cmd *Cmd
	for i := range Cmds {
		if Cmds[i].Name == word {
			cmd = &Cmds[i]
		}
	}
	// Slash lines are never sent to the model; unknown ones just warn.
	if cmd == nil {
		if p, err := c.customPrompt(word, arg); err == nil {
			return true, false, p
		} else if !errors.Is(err, errNoCustom) {
			c.ui.Printf("❌ %s: %v", word, err)
			return true, false, ""
		}
		c.ui.Printf("⚠️  unknown command %s (try /help)", word)
		return true, false, ""
	}

	switch cmd.Name {
	case "/quit":
		return true, true, ""
	case "/help":
		for _, cc := range Cmds {
			c.ui.Printf("  %-8s %-16s %s", cc.Name, cc.Args, cc.Desc)
		}
		for _, cc := range c.CustomCmds() {
			c.ui.Printf("  %-25s %s", cc.Name, cc.Desc)
		}
	case "/new":
		c.cmdNew(ag)
	case "/compact": // interactive drivers intercept this before Dispatch
		c.ui.Printf("⚠️  /compact is unavailable in this mode")
	case "/resume":
		c.cmdResume(ag, arg)
	case "/status":
		c.cmdStatus(ag)
	case "/model":
		c.cmdModel(ag, arg)
	case "/effort":
		c.cmdEffort(ag, arg)
	}
	return true, false, ""
}

// cmdResume switches the live agent to another session; no arg lists them.
func (c *Commands) cmdResume(ag *agent.Agent, arg string) {
	if arg == "" {
		rows, err := session.List(c.cfg.Cwd)
		if err != nil || len(rows) == 0 {
			c.ui.Printf("📭 no sessions")
			return
		}
		for _, r := range rows {
			marker := ' '
			if strings.HasPrefix(c.cfg.SessionID, r.ID) || strings.HasPrefix(r.ID, c.cfg.SessionID) {
				marker = '*'
			}
			c.ui.Printf("%c %-8.8s %-16s %s", marker, r.ID, r.When, r.Title)
		}
		return
	}
	prevID := c.cfg.SessionID
	next, resumed, err := session.Resume(arg, c.cfg)
	if err != nil {
		c.cfg.SessionID = prevID
		c.ui.Printf("❌ %v", err)
		return
	}
	old := ag.Sess
	old.Close()
	if old.Items == 0 && old.Path != next.Path {
		os.Remove(old.Path)
	}
	*ag.Sess = *next
	ag.History = resumed
	if next.Model != "" {
		c.cfg.Model = next.Model
	}
	if next.Effort != "" {
		c.cfg.Effort = next.Effort
	}
	c.ui.Printf("↩️  resumed %.8s (%d items)", c.cfg.SessionID, next.Items)
	ag.Replay()
	c.CtxUsed(next.Ctx)
}

func (c *Commands) cmdStatus(ag *agent.Agent) {
	c.ui.Printf("🧌 orc %s · session %.8s (%d items)", config.Version, c.cfg.SessionID, ag.Sess.Items)
	c.ui.Printf("📄 %s", ag.Sess.Path)
	c.ui.Printf("🤖 %s (%s effort) · provider %s", c.cfg.Model, c.cfg.Effort, c.prov.Name())
	if c.ctxUsed > 0 {
		s := fmt.Sprintf("📊 ctx %d tokens", c.ctxUsed)
		if win := c.modelCtxWindow(); win > 0 {
			s += fmt.Sprintf(" of %dk (%d%%)", win/1000, (c.ctxUsed*100+win-1)/win)
		}
		c.ui.Printf("%s", s)
	}
	cwd := c.cfg.Cwd
	if cwd == "" {
		cwd, _ = os.Getwd()
	}
	if branch := gitBranch(cwd); branch != "" {
		cwd += fmt.Sprintf(" (%s)", branch)
	}
	c.ui.Printf("📁 %s", cwd)
}

// Sessions lists resumable sessions for the menu, current one excluded.
func (c *Commands) Sessions() []session.Info {
	rows, _ := session.List(c.cfg.Cwd)
	out := rows[:0]
	for _, r := range rows {
		if !strings.HasPrefix(c.cfg.SessionID, r.ID) && !strings.HasPrefix(r.ID, c.cfg.SessionID) {
			out = append(out, r)
		}
	}
	return out
}

func (c *Commands) cmdModel(ag *agent.Agent, arg string) {
	models := c.Models()
	if arg == "" {
		c.ui.Printf("🤖 model %s (%s effort)", c.cfg.Model, c.cfg.Effort)
		for _, m := range models {
			marker := ' '
			if m.Slug == c.cfg.Model {
				marker = '*'
			}
			c.ui.Printf("  %c %-22s %s", marker, m.Slug, m.Description)
		}
		return
	}
	known := models == nil // no list: trust the user
	for _, m := range models {
		if m.Slug == arg {
			known = true
		}
	}
	c.cfg.Model = arg
	note := ""
	if !known {
		note = " (not in provider's model list)"
	}
	c.ui.Printf("✅ model set to %s%s", arg, note)
	ag.Sess.SetCfg(c.cfg)
	c.saveDefaults()
	c.StatusUpdate()
}

// saveDefaults makes the current model/effort the defaults for new sessions.
func (c *Commands) saveDefaults() {
	config.SaveSettings(config.Settings{Model: c.cfg.Model, Effort: c.cfg.Effort})
}

func (c *Commands) cmdEffort(ag *agent.Agent, arg string) {
	if arg == "" {
		c.ui.Printf("🧠 effort %s", c.cfg.Effort)
		return
	}
	if arg != "low" && arg != "medium" && arg != "high" {
		c.ui.Printf("⚠️  effort must be low, medium, or high")
		return
	}
	c.cfg.Effort = arg
	c.ui.Printf("✅ effort set to %s", arg)
	ag.Sess.SetCfg(c.cfg)
	c.saveDefaults()
	c.StatusUpdate()
}
